onnx/defs/controlflow/defs.cc

```diff
@@ -0,0 +1,215 @@
+    ) {
+      %my_local = Add(%a, %b)
+      %b_out = Sub(%a, %b)
+      %keepgoing_out = Greater(%my_local, %b_out)
+      %user_defined_vals = Add(%b, %b)
+      return %keepgoing_out, %b_out, %user_defined_vals
+    }
+
+*Sample equivalent C code*
+
+    {
+      /* User-defined code (enclosing scope) */
+      int a = 3, b = 6;
+      bool keepgoing = true; // Analogous to input cond
+      /* End user-defined code */
+
+      /* Implicitly-defined code */
+      const int max_trip_count = 10; // Analogous to input M
+      int user_defined_vals[]; // Imagine this is resizable
+      /* End implicitly-defined code */
+      for (int i=0; i < max_trip_count && keepgoing; ++i) {
+        /* User-defined code (loop body) */
+        int my_local = a + b; // Reading values in the enclosing scope is fine
+        b = a - b; // writes fine if we specify b as a loop-carried dependency
+        keepgoing = my_local > b; // keepgoing is a loop-carried dependency
+        user_defined_vals[i] = b + b;
+        /* End user-defined code */
+      }
+      // my_local = 123; // Can't do this. my_local was defined in the the body
+
+      // These below values are live-out from the loop and therefore accessible
+      b_out; user_defined_vals; keepgoing_out;
+    }
+
+There are several things of note in this code snippet:
+
+1) Values from the enclosing scope (i.e. variable a here) are in scope and can
+   be referenced in the inputs of the loop.
+2) Any variables which you wish to make available in the enclosing scope (i.e.
+   the variables b and keepgoing) must be declared as either loop-carried
+   dependencies (both at the op inputs and output and at the body net input and
+   output) or scan_outputs.
+3) Values created in the body cannot be accessed in the enclosing scope.
+
+Note that the semantics of this op support "diagonal" or "wavefront" execution.
+(See Step 3 here for an example:
+https://devblogs.nvidia.com/optimizing-recurrent-neural-networks-cudnn-5/).
+Frontends should emit multi-layer RNNs as a series of While operators (with
+time being the inner looping dimension), with each successive layer consuming
+the scan_outputs from the previous layer, possibly going through several
+point-wise operators (e.g. dropout, residual connections, linear layer).
+Concretely, the (possibly transformed) scan_outputs are referenced by the
+subsequent layer as a LoopIndexTensor operating on a value in scope, not
+necessarily a loop-carried dependency. Backends can recognize this pattern and
+are permitted to schedule the execution of the multi-layer network in a
+pipelined/"wavefront" fashion.
+
+)DOC")
+    .Input(
+        0,
+        "M",
+        "A maximum trip-count for the loop specified at runtime. Optional."
+        " pass empty string to skip.",
+        "I")
+    .Input(
+        1,
+        "cond",
+        "A boolean termination condition. Pass empty string to skip.",
+        "B")
+    .Input(
+        2,
+        "v_initial",
```

Comment:
We have Variadic input and Variadic output, how does the backend distinguish them? Shall we add some attributes to help us figure out?